pkg/controller/podautoscaler/horizontal.go

```diff
@@ -520,6 +486,57 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 	return a.updateStatusIfNeeded(hpaStatusOriginal, hpa)
 }
 
+// normalizeDesiredReplicas takes the metrics desired replicas value and normalizes it based on the appropriate conditions (i.e. < maxReplicas, >
+// minReplicas, etc...)
+func (a *HorizontalController) normalizeDesiredReplicas(hpa *autoscalingv2.HorizontalPodAutoscaler, currentReplicas int32, prenormalizedDesiredReplicas int32) (int32, error) {
+	var minReplicas int32
+	if hpa.Spec.MinReplicas != nil {
+		minReplicas = *hpa.Spec.MinReplicas
+	} else {
+		minReplicas = 0
+	}
+
+	desiredReplicas, err := convertDesiredReplicasWithRules(currentReplicas, prenormalizedDesiredReplicas, minReplicas, hpa.Spec.MaxReplicas)
+	if err != nil {
+		return -1, err
+	}
+
+	if desiredReplicas < prenormalizedDesiredReplicas {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooManyReplicas", "the desired replica count is above the maximum number of allowed replicas")
+	} else if desiredReplicas > prenormalizedDesiredReplicas {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionTrue, "TooFewReplicas", "the desired replica count is below the minimum number of allowed replicas")
+	} else {
+		setCondition(hpa, autoscalingv2.ScalingLimited, v1.ConditionFalse, "DesiredWithinRange", "the desired replica count is within the acceptable range")
+	}
+
+	return desiredReplicas, nil
+}
+
+func convertDesiredReplicasWithRules(currentReplicas, desiredReplicas, hpaMinReplicas, hpaMaxReplicas int32) (int32, error) {
+	// Do not upscale too much to prevent incorrect rapid increase of the number of master replicas caused by
+	// bogus CPU usage report from heapster/kubelet (like in issue #32304).
+	scaleUpLimit := calculateScaleUpLimit(currentReplicas)
+
+	minimumAllowedReplicas := int32(math.Max(float64(hpaMinReplicas), 1.0))
```

Comment:
don't use `math.Max` and `math.Min` on ints.  Converting to and from floats can cause issues, and is generally not appropriate here.  Just use an if statement.

(FWIW, I think that Go should have `Max` and `Min` functions for ints as well)

Reply:
Ah great to know! Agreed Go should have `Max/Min` functions for ints...